[PM-7343] AnonLayoutComponent Implementation Groundwork

rr-bw opened the pull request:
## Type of change


- New feature development


## Objective

https://github.com/bitwarden/clients/pull/6424 dealt with creating the Auth-owned AnonLayoutComponent in Storybook, but did not involve any implementation in clients.

This PR lays the groundwork for actually implementing the AnonLayoutComponent in clients (starting with Web), and provides accompanying [Storybook docs](https://62a88a6de5b807fa98886113-znknkdmxqq.chromatic.com/?path=/docs/auth-anon-layout--docs).

## Background

When creating the AnonLayoutComponent, we decided that it should be implemented via routable composition (rather than by every component using the AnonLayoutComponent directly in the component's template). Routable composition gives us the advantages of nested routes in Angular. The background for that conversation can be found in [this comment](https://github.com/bitwarden/clients/pull/6424#pullrequestreview-1679952327) and this [follow-up comment](https://github.com/bitwarden/clients/pull/6424#issuecomment-1879311281).

## Usage

This PR creates an additional [AnonLayoutWrapperComponent](https://github.com/bitwarden/clients/pull/8585/files#diff-cab22f830360037c317d025d439f8af083c887555045d20433c449050839ccc5). Inside this wrapper component exists the AnonLayoutComponent (`<auth-anon-layout>`) along with the router outlets.

For clarity:
- AnonLayoutComponent = the Auth-owned library component - `<auth-anon-layout>`
- AnonLayoutWrapperComponent = the client-specific wrapper component to be used in the client routing module

```html
<!-- File: anon-layout-wrapper.component.html -->

<auth-anon-layout>
  <router-outlet></router-outlet>
  <router-outlet slot="secondary" name="secondary"></router-outlet>
</auth-anon-layout>
```

To implement, the developer uses the AnonLayoutWrapperComponent in the router (`oss-routing.module.ts`) like so:

```javascript
{
    path: "",
    component: AnonLayoutWrapperComponent, // Wrapper component
    children: [
      {
        path: "sample-route", // replace with your route
        children: [
          {
            path: "",
            component: MyPrimaryComponent, // replace with your component
          },
          {
            path: "",
            component: MySecondaryComponent, // replace with your component (or remove this secondary outlet object entirely if not needed)
            outlet: "secondary",
          },
        ],
        data: {
          pageTitle: "logIn", // example of a translation key from messages.json
          pageSubtitle: "loginWithMasterPassword", // example of a translation key from messages.json
          pageIcon: LockIcon, // example of an icon to pass in
        },
      },
    ],
  },
```

Note: this implementation is similar to our Secrets Manager routing module, where the [LayoutComponent](https://github.com/bitwarden/clients/blob/main/bitwarden_license/bit-web/src/app/secrets-manager/layout/layout.component.html) in Bit Web uses the `<bit-layout>` component from the Component Library. The LayoutComponent is then implemented via routable composition in [sm-routing.module.ts](https://github.com/bitwarden/clients/blob/45dc244cc41755a0442b9d40908fce37511c9662/bitwarden_license/bit-web/src/app/secrets-manager/sm-routing.module.ts#L31)

## Example Implementation 🚨

To see an example of how this would be implemented, I've created another temporary branch that you can build to see it in action: https://github.com/bitwarden/clients/pull/8984

trmartin4 commented:
This looks great!  I can't comment on the technical side of the implementation, but I wanted to ask about a slight tweak in the documentation and the PR description.  I'd suggest avoiding referring to the `AnonLayoutComponent` as a "CL component".  It's not technically owned by the CL team, as it's a shared component owned by Auth.  Perhaps instead call it an "Auth-owned library component"?

rr-bw commented:
> This looks great! I can't comment on the technical side of the implementation, but I wanted to ask about a slight tweak in the documentation and the PR description. I'd suggest avoiding referring to the `AnonLayoutComponent` as a "CL component". It's not technically owned by the CL team, as it's a shared component owned by Auth. Perhaps instead call it an "Auth-owned library component"?

Great catch @trmartin4. I will update that.

jlf0dev commented on libs/auth/src/angular/anon-layout/anon-layout.mdx:
```diff
@@ -0,0 +1,59 @@
+template directly. Instead the AnonLayoutComponent is implemented via routable composition, which
+gives us the advantages of nested routes in Angular.
+
+To allow for routable composition, Web has an AnonLayoutWrapperComponent which embeds the Auth-owned
+AnonLayoutComponent.
```
How does this work for desktop and browser? Will they have wrapper components as well?

jlf0dev reviewed (commented):
This is awesome! Thank you for the details you've included in the documentation, I'm sure that will pay dividends in the future.

Hinton commented on apps/web/src/app/shared/shared.module.ts:
```diff
@@ -80,6 +81,7 @@ import "./locales";
     TabsModule,
     ToggleGroupModule,
     TypographyModule,
+    AnonLayoutComponent,
```
I don't think this qualifies to be added here.
```
* This NgModule should contain the most basic shared directives, pipes, and components. They
* should be widely used by other modules to be considered for adding to this module. If in doubt
* do not add to this module.
```

Hinton commented on apps/web/src/app/oss-routing.module.ts:
```diff
@@ -188,6 +189,25 @@ const routes: Routes = [
     ],
   },
+
+  // TODO: remove this example
+  {
+    path: "",
+    component: AnonLayoutWrapperComponent,
+    children: [
+      {
+        path: "sample-route",
+        component: LoginComponent, // replace with your component
+      },
+      {
+        path: "",
+        component: HintComponent, // just an example that shows secondary content. Replace with your component (or remove this secondary outlet entirely if not needed)
+        outlet: "secondary",
+      },
+    ],
+  },
+  // TODO: remove this example
```
Comment to ensure this gets removed

rr-bw replied:
Removed. However I opened a [temporary draft PR](https://github.com/bitwarden/clients/pull/8984) with this route added in case you want to test visually for yourself. Will close after.

rr-bw commented:
@Hinton I updated the AnonLayoutWrapperComponent to allow passing down the `pageTitle`, `pageSubtitle`, and `pageIcon` from the router. For the `pageIcon`, unfortunately this meant reintroducing a [switch statement](https://github.com/bitwarden/clients/pull/8585/files#diff-cfa739dce06ebf3fb10c53f830d3c75aa1a3500336e40ef410bd01f5eea8fd04R32-R38) back into the auth AnonLayoutComponent. Let me know if you think there is a better way to handle the icon.

One other question, I noticed that my `tw` classes aren't getting set in development. For example, the classes `tw-gap-9 tw-pt-14 tw-fill-primary-500 tw-mb-8` are all used in the AnonLayoutComponent, but don't actually affect styling until I add them arbitrarily to another Web component like `login.component.html`. Am I doing something wrong with imports?

Hinton commented:
@rr-bw I would lift up the switch statement into the wrapper. The wrapper parses the url routes, not the base component.

Did we test if passing in a icon in the routes file would work?

rr-bw commented:
@Hinton Thanks, I moved the icon to the router and fixed the tailwind issues.

rr-bw commented:
@Hinton @jlf0dev This is ready for a final review. A few comments and one question:
- I've moved the `AnonLayoutWrapperComponent` to `libs` as well so it can be used across clients.
- I've updated storybook docs.
- If you want to see this work in action, I've created [temporary PR](https://github.com/bitwarden/clients/pull/8984) with a `sample-route` that you can build locally to view.
- One question: The `AnonLayoutComponent` is in fact responsive (in storybook), but for it to truly be responsive in Web, we'll have to wait until we deal with the `<meta name="viewport" content="width=1010" />` in `index.html`, correct?

Hinton commented on libs/auth/src/angular/anon-layout/anon-layout-wrapper.component.ts:
I consider routable components to be application specific. While you can ensure each application is routed identically it's not something we currently do.

rr-bw replied:
Makes sense. I moved it back to web.

Hinton commented on libs/auth/src/angular/anon-layout/anon-layout-wrapper.component.ts:
```diff
@@ -0,0 +1,36 @@
+    this.pageIcon = this.route.snapshot.firstChild.data["pageIcon"]; // don't translate
+  }
+
+  ngOnInit() {
+    document.body.classList.add("layout_frontend");
+  }
+
+  ngOnDestroy() {
+    document.body.classList.remove("layout_frontend");
+  }
```
Web specific.
